x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/string/Left.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.expression.function.scalar.string;
+
+import org.apache.lucene.util.BytesRef;
+import org.apache.lucene.util.UnicodeUtil;
+import org.elasticsearch.compute.ann.Evaluator;
+import org.elasticsearch.compute.operator.EvalOperator;
+import org.elasticsearch.xpack.esql.planner.Mappable;
+import org.elasticsearch.xpack.ql.expression.Expression;
+import org.elasticsearch.xpack.ql.expression.function.scalar.ScalarFunction;
+import org.elasticsearch.xpack.ql.expression.gen.script.ScriptTemplate;
+import org.elasticsearch.xpack.ql.tree.NodeInfo;
+import org.elasticsearch.xpack.ql.tree.Source;
+import org.elasticsearch.xpack.ql.type.DataType;
+import org.elasticsearch.xpack.ql.type.DataTypes;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.ParamOrdinal.FIRST;
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.ParamOrdinal.SECOND;
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.ParamOrdinal.THIRD;
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.isInteger;
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.isString;
+
+/**
+ * left(foo, len) is a alias that substring(foo, 0, len)
+ */
+public class Left extends ScalarFunction implements Mappable {
+
+    private final Expression str;
+
+    private final Expression length;
+
+    public Left(Source source, Expression str, Expression length) {
+        super(source, Arrays.asList(str, length));
+        this.str = str;
+        this.length = length;
+    }
+
+    @Evaluator
+    static BytesRef process(BytesRef str, int length) {
+        if (str.length == 0) {
+            return str;
+        }
+        if (length < 0) {
+            throw new IllegalArgumentException("Length parameter cannot be negative, found [" + length + "]");
+        }
+
+        int codePointCount = UnicodeUtil.codePointCount(str);
+        int indexEnd = Math.min(codePointCount, length);
+        String s = str.utf8ToString();
+        return new BytesRef(s.substring(0, indexEnd));
```

Comment:
It'd be lovely if you could do this without allocating on ever iteration. I *think* it's possible with something like:
```
static BytesRef process(@Fixed BytesRef out, @Fixed UTF8CodePoint cp, BytesRef str, int length) {
  ...
  result.bytes = str.bytes;
  result.offset = str.offset;
  result.length = str.length;
  for (int i = 0; i < length; i++) {
    cp = UnicodeUtil.codePointAt(result.bytes, result.start, cp);
    result.start += cp.numBytes;
    result.length -= cp.numBytes;
  }
  result.length = Math.max(0, result.length);
  return result;
}
```

At least, I think something like that'd work. Actually, I think I wrote `right` instead of `left`, but something like that.

With yours we have to convert the whole thing into a utf16 string and back again. Also, I'm a bit worried about places where utf-16 characters aren't a single code point. I think this'll cut them up differently.

Reply:
Hmmm. It looks like `Substring` has the same problem....

Reply:
Another option, I'm not sure if it's worth it, is to just return the `SubstringEvaluator` here instead.

Reply:
It's good that just return SubstringEvaluator(str, 0, length)

Reply:
> Hmmm. It looks like `Substring` has the same problem....

I just modify the string in place at first?

Reply:
What's wrong with my opertion?
I check out a branch named 'left-function'  at local and commit some code.
Then I merge the code from elastic repo into local branch and push rebasely to my forked repo.
But why is there such a large difference in the pull request?